packages/portfolio-contract/tools/plan-solve.ts

```diff
@@ -388,6 +392,130 @@ export const solveRebalance = async (
   return { flows, detail: { solution } };
 };
 
+/**
+ * Compute partial order from solved flows and initial supplies.
+ *
+ * For each node, we track available supply (initial + completed inflows).
+ * An outflow step depends on the minimum set of inflow steps needed to
+ * provide sufficient supply.
+ *
+ * This approach correctly handles cases like:
+ * - Multiple operations to/from the same node with separate dependencies
+ * - Initial supplies that satisfy some outflows immediately
+ * - Fan-out patterns where multiple outflows depend on the same inflow
+ *
+ * @param flows - solved flows from the LP solver
+ * @param initialSupplies - starting balances at each node
+ * @returns prioritized flows in execution order and partial order constraints
+ */
+export const computePartialOrder = (
+  flows: SolvedEdgeFlow[],
+  initialSupplies: Map<string, number>,
+): { prioritized: SolvedEdgeFlow[]; order?: StepOrder } => {
+  // Return immediately if there's nothing to do.
+  if (flows.length === 0) return { prioritized: [] };
+
+  /** Available supply at each node */
+  const available = new Map(initialSupplies);
+
+  /** Remaining initial supply at each node (for dependency calculation) */
+  const initialRemaining = new Map(initialSupplies);
+
+  /** Which step indices have contributed to each node's balance */
+  const inflows = new Map<
+    AssetPlaceRef,
+    { flowIdx: number; amount: number }[]
+  >();
+
+  const order: StepOrder = [];
+  const prioritized: SolvedEdgeFlow[] = [];
+  const scheduled = new Set<number>();
+
+  // Maintain last chosen originating chain to group sequential operations.
+  let lastChain: string | undefined;
+
+  while (scheduled.size < flows.length) {
+    /** Flows that can execute with current available supply */
+    const candidates: { idx: number; flow: SolvedEdgeFlow }[] = [];
+    for (let idx = 0; idx < flows.length; idx += 1) {
```

Comment:
`.filter(...)` isn't handy here?

Reply:
No, because we need the original indices.